src/irmin/tree.ml

```diff
@@ -622,12 +678,40 @@ module Make (P : Private.S) = struct
             | Some x, Some y -> if equal_node x y then True else False
             | _ -> Maybe)
 
+    let is_updatemap_empty v um =
+      let any_add =
+        StepMap.to_seq um
+        |> seq_exists (function _, Remove -> false | _, Add _ -> true)
+      in
+      if any_add then false
+      else
+        let val_is_empty = P.Node.Val.is_empty v in
+        if val_is_empty then true
+        else
+          let remove_count = StepMap.cardinal um in
+          if (not val_is_empty) && remove_count = 0 then false
+          else (
+            (* This is expensive from this point but there is no alternative *)
+            cnt.node_val_list <- cnt.node_val_list + 1;
+            let entries = P.Node.Val.list v in
+            if List.length entries > remove_count then false
```

Comment:
Unfolding this would let you short-circuit as soon as > remove_count entries are seen, which is going to be the common case after flattening.

Reply:
Not sure I understood you correctly. Are you suggesting that testing the length won't be part of this function after implementing `P.Node.Val.seq`? If so, I agree.

A second optimisation would be to add `P.Node.Val.length` too, to keep this check and avoid iterating at all on the (i)node.

Reply:
As it stands, `P.Node.Val.list` may create a list of 1 million entries. Suppose `remove_count = 5` here, we still traverse the entire list to get its length even though we're guaranteed to return `false` after seeing 6 nodes. If you write a little recursive function by hand here, it's possible to abort that traversal early.

Reply:
I see! 

I didn't know (or forgot) that `List.length` required traversing the list.

Reply:
Fixed: https://github.com/mirage/irmin/pull/1289/commits/03506ee5a9109f5a1b07338ff537f1cf219dc487